File: frontend/types/order.ts
import { z } from "zod";
import { GasTypeSchema, OrderModeSchema } from "./common";

export const OrderSchema = z
  .object({
    id: z.string(),
    customer_id: z.string(),
    system_id: z.string(),
    delivered_at: z.string(),
    created_at: z.string(),
    updated_at: z.string().nullish(),
    order_mode: OrderModeSchema.optional(),
    gas_type: GasTypeSchema,
    cylinders_installed: z.number(),
    cylinders_received: z.number(),
    price_total: z.number(),
    paid_amount: z.number().optional(),
    debt_cash: z.number().optional().default(0),
    debt_cylinders_12: z.number().optional().default(0),
    debt_cylinders_48: z.number().optional().default(0),
    applied_credit: z.number().optional().nullish(),
    money_balance_before: z.number().optional().nullish(),
    money_balance_after: z.number().optional().nullish(),
    cyl_balance_before: z.record(z.string(), z.number()).optional().nullish(),
    cyl_balance_after: z.record(z.string(), z.number()).optional().nullish(),
    note: z.string().nullish(),
    is_deleted: z.boolean().optional(),
  })
  .passthrough();
export type Order = z.infer<typeof OrderSchema>;

export const OrderCreateInputSchema = z.object({
  customer_id: z.string(),
  system_id: z.string().optional(),
  delivered_at: z.string().optional(),
  order_mode: OrderModeSchema.optional(),
  gas_type: GasTypeSchema,
  cylinders_installed: z.number(),
  cylinders_received: z.number(),
  price_total: z.number(),
  paid_amount: z.number().optional(),
  debt_cash: z.number().optional(),
  debt_cylinders_12: z.number().optional(),
  debt_cylinders_48: z.number().optional(),
  note: z.string().nullish().optional(),
  request_id: z.string().optional(),
});
export type OrderCreateInput = z.infer<typeof OrderCreateInputSchema>;

export const OrderUpdateInputSchema = OrderCreateInputSchema.partial();
export type OrderUpdateInput = z.infer<typeof OrderUpdateInputSchema>;

export const CollectionCreateInputSchema = z.object({
  customer_id: z.string(),
  action_type: z.enum(["payment", "payout", "return"]),
  amount_money: z.number().optional(),
  qty_12kg: z.number().optional(),
  qty_48kg: z.number().optional(),
  debt_cash: z.number().optional(),
  debt_cylinders_12: z.number().optional(),
  debt_cylinders_48: z.number().optional(),
  system_id: z.string().nullish().optional(),
  effective_at: z.string().optional(),
  note: z.string().nullish().optional(),
  request_id: z.string().optional(),
});
export type CollectionCreateInput = z.infer<typeof CollectionCreateInputSchema>;

export const CollectionUpdateInputSchema = CollectionCreateInputSchema.partial();
export type CollectionUpdateInput = z.infer<typeof CollectionUpdateInputSchema>;

export const CollectionEventSchema = z
  .object({
    id: z.string(),
    customer_id: z.string(),
    action_type: z.enum(["payment", "payout", "return"]),
    amount_money: z.number().nullish(),
    qty_12kg: z.number().nullish(),
    qty_48kg: z.number().nullish(),
    debt_cash: z.number().nullish(),
    debt_cylinders_12: z.number().nullish(),
    debt_cylinders_48: z.number().nullish(),
    system_id: z.string().nullish(),
    created_at: z.string(),
    effective_at: z.string().nullish(),
    note: z.string().nullish(),
    is_deleted: z.boolean().optional(),
  })
  .passthrough();
export type CollectionEvent = z.infer<typeof CollectionEventSchema>;

export const OrderImpactSchema = z.object({
  gross_paid: z.number(),
  applied_credit: z.number(),
  unpaid: z.number(),
  new_balance: z.number(),
  cyl_balance_before: z.record(z.string(), z.number()).optional(),
  cyl_balance_after: z.record(z.string(), z.number()).optional(),
});
export type OrderImpact = z.infer<typeof OrderImpactSchema>;

File: backend/app/schemas/order.py
from __future__ import annotations

from datetime import datetime
from typing import Literal, Optional

from pydantic import field_validator
from sqlmodel import SQLModel

from .common import GasType, OrderMode, _non_negative


class OrderCreate(SQLModel):
  customer_id: str
  system_id: Optional[str] = None
  happened_at: Optional[datetime] = None
  order_mode: OrderMode = "replacement"
  gas_type: GasType
  cylinders_installed: int
  cylinders_received: int
  price_total: int
  paid_amount: Optional[int] = None
  debt_cash: int = 0
  debt_cylinders_12: int = 0
  debt_cylinders_48: int = 0
  note: Optional[str] = None
  request_id: Optional[str] = None

  @field_validator("cylinders_installed", "cylinders_received", "price_total", "paid_amount")
  @classmethod
  def _validate_non_negative(cls, value: Optional[int], info) -> Optional[int]:
    return _non_negative(value, info.field_name)


class OrderUpdate(SQLModel):
  customer_id: Optional[str] = None
  system_id: Optional[str] = None
  happened_at: Optional[datetime] = None
  order_mode: Optional[OrderMode] = None
  gas_type: Optional[GasType] = None
  cylinders_installed: Optional[int] = None
  cylinders_received: Optional[int] = None
  price_total: Optional[int] = None
  paid_amount: Optional[int] = None
  debt_cash: Optional[int] = None
  debt_cylinders_12: Optional[int] = None
  debt_cylinders_48: Optional[int] = None
  note: Optional[str] = None

  @field_validator("cylinders_installed", "cylinders_received", "price_total", "paid_amount")
  @classmethod
  def _validate_non_negative(cls, value: Optional[int], info) -> Optional[int]:
    return _non_negative(value, info.field_name)


class OrderOut(SQLModel):
  id: str
  customer_id: str
  system_id: Optional[str] = None
  delivered_at: datetime
  created_at: datetime
  updated_at: Optional[datetime] = None
  order_mode: OrderMode
  gas_type: GasType
  cylinders_installed: int
  cylinders_received: int
  price_total: int
  paid_amount: int
  debt_cash: int = 0
  debt_cylinders_12: int = 0
  debt_cylinders_48: int = 0
  note: Optional[str] = None
  money_balance_before: Optional[int] = None
  money_balance_after: Optional[int] = None
  cyl_balance_before: Optional[dict[str, int]] = None
  cyl_balance_after: Optional[dict[str, int]] = None
  is_deleted: bool = False


class CollectionCreate(SQLModel):
  customer_id: str
  action_type: Literal["payment", "payout", "return"]
  amount_money: Optional[int] = None
  qty_12kg: Optional[int] = None
  qty_48kg: Optional[int] = None
  debt_cash: int = 0
  debt_cylinders_12: int = 0
  debt_cylinders_48: int = 0
  system_id: Optional[str] = None
  happened_at: Optional[datetime] = None
  note: Optional[str] = None
  request_id: Optional[str] = None

  @field_validator("amount_money", "qty_12kg", "qty_48kg")
  @classmethod
  def _validate_non_negative(cls, value: Optional[int], info) -> Optional[int]:
    return _non_negative(value, info.field_name)

  @field_validator("system_id")
  @classmethod
  def _validate_system_id(cls, value: Optional[str]) -> Optional[str]:
    if value is not None:
      raise ValueError("system_not_allowed")
    return value


class CollectionUpdate(SQLModel):
  action_type: Optional[Literal["payment", "payout", "return"]] = None
  amount_money: Optional[int] = None
  qty_12kg: Optional[int] = None
  qty_48kg: Optional[int] = None
  debt_cash: Optional[int] = None
  debt_cylinders_12: Optional[int] = None
  debt_cylinders_48: Optional[int] = None
  system_id: Optional[str] = None
  happened_at: Optional[datetime] = None
  note: Optional[str] = None

  @field_validator("amount_money", "qty_12kg", "qty_48kg")
  @classmethod
  def _validate_non_negative(cls, value: Optional[int], info) -> Optional[int]:
    return _non_negative(value, info.field_name)

  @field_validator("system_id")
  @classmethod
  def _validate_system_id(cls, value: Optional[str]) -> Optional[str]:
    if value is not None:
      raise ValueError("system_not_allowed")
    return value


class CollectionEvent(SQLModel):
  id: str
  customer_id: str
  action_type: Literal["payment", "payout", "return"]
  amount_money: Optional[int] = None
  qty_12kg: Optional[int] = None
  qty_48kg: Optional[int] = None
  debt_cash: int = 0
  debt_cylinders_12: int = 0
  debt_cylinders_48: int = 0
  system_id: Optional[str] = None
  created_at: datetime
  effective_at: datetime
  note: Optional[str] = None
  is_deleted: bool = False
